src/mpi/coll/algorithms/recexchalgo/recexchalgo.c

```diff
@@ -276,7 +276,7 @@ int MPII_Recexchalgo_reverse_digits_step2(int rank, int comm_size, int k)
     int i, T, rem, power, step2rank, step2_reverse_rank = 0;
     int pofk = 1, log_pofk = 0;
     int *digit, *digit_reverse;
-    int mpi_errno = MPI_SUCCESS;
+    int mpi_errno ATTRIBUTE((unused)) = MPI_SUCCESS;
```

Comment:
I guess that would also get rid of the problem, but that assertion is just...terrible...

I'd feel gross touching that code in any way but to delete it.

[checks `git blame`] ... oh God ...

Reply:
I start to regard `ATTRIBUTE((unused))` as last resort warning fix, since it essentially just a suppression without reasons. The ideal solution would be rid of the variable or move the declarations inside macro guard or branches. For boiler-plate code such as mpi_errno, I think `MPIR_Assert` is a second last solution. For one, we already use it every where; for two, the boiler plate is there for "future" code expansions/changes, so `MPIR_Assert` can be regarded as a placeholder.

Reply:
> [checks `git blame`] ... oh God ...

Almost missed this 😆 😆 😆

Reply:
BTW, moving the assertion won't help in the case where you're building without error checking. In that case, `MPIR_Assert` becomes nothing (which is the case I was trying to avoid here).

Reply:
> BTW, moving the assertion won't help in the case where you're building without error checking. In that case, `MPIR_Assert` becomes nothing (which is the case I was trying to avoid here).

Yeah, I realize that. It seems you are trying `--enable-strict` with `--enable-fast=all` at the same time? Is this test necessary?

Reply:
I always build with `--enable-strict`, regardless of everything else. What's the problem with catching and removing warnings? 😄

Reply:
> I always build with `--enable-strict`, regardless of everything else. What's the problem with catching and removing warnings? 😄

Because in my view, `--enable-strict` and `--enable-fast` is of opposite intention. The former is for correctness, the later is for production. Just like we would never enable sanitizers for production, even though we could.

Reply:
Adding an attribute for an unused variable has no impact on performance. So I don't see how the goals are orthogonal.

Reply:
By https://github.com/pmodels/mpich/pull/4751#discussion_r475749316 my question is why would any one do `--enable-strict` and `--enable-fast` at the same time? and if they do, why? In your case, I know you do, but only because you "always build with --enable-strict, regardless of everything else"? So if you change this "always", and don't add `--enable-strict` for tests with `--enable-fast`, or ignore the warnings when you do, will the issues go away?

Reply:
My push back is, to my impression, we use `MPIR_Assert` to suppress warnings quite often, so if from now on, we are not allowed to do that and should use `attribute((unused))` in stead, it'll be a wide impact decision. So I think it requires a separate PR discussion  and a conclusive decision of some sort.

Reply:
I didn't say you can't use assertions. In cases where that is effective, that's fine. In this case, it doesn't solve the problem.

One of the major benefits of solving this is that `-Werror` now works so we could have a more reliable warning checker in Jenkins.

Reply:
> I didn't say you can't use assertions. In cases where that is effective, that's fine. In this case, it doesn't solve the problem.

In all cases, --enable-fast=NDEBUG will render MPIR_Assert disappear, right? So "this case" is all cases that using `MPIR_Assert` to suppress warnings.

Also, is `__attribute__` part of standard? It looks like it is compiler specific, and the newer C++ standard is introducing different syntax.

Reply:
> One of the major benefits of solving this is that -Werror now works so we could have a more reliable warning checker in Jenkins.

Did you solve all the unused goto label warnings from `icc`?

Reply:
Fair point on that. I was just focused on GCC this time around.

Reply:
If you are interested, in https://github.com/pmodels/mpich/pull/4737/commits, I did manage to fix all the goto label warnings. That PR become to extensive, so probably won't get considered until 4.0.